pkg/status/status.go

```diff
@@ -0,0 +1,219 @@
+	return p2p.ProtocolSpec{
+		Name:    protocolName,
+		Version: protocolVersion,
+		StreamSpecs: []p2p.StreamSpec{{
+			Name:    streamName,
+			Handler: s.handler,
+		}},
+	}
+}
+
+// requestStatusSnapshot sends request for status snapshot to the peer.
+func (s *Service) requestStatusSnapshot(ctx context.Context, peer swarm.Address) (*pb.Snapshot, error) {
+	stream, err := s.streamer.NewStream(ctx, peer, nil, protocolName, protocolVersion, streamName)
+	if err != nil {
+		return nil, fmt.Errorf("new stream: %w", err)
+	}
+	defer func() {
+		go stream.FullClose()
+	}()
+
+	w, r := protobuf.NewWriterAndReader(stream)
+
+	if err := w.WriteMsgWithContext(ctx, new(pb.Get)); err != nil {
+		return nil, fmt.Errorf("write message failed: %w", err)
+	}
+
+	ss := new(pb.Snapshot)
+	if err := r.ReadMsgWithContext(ctx, ss); err != nil {
+		if errors.Is(err, io.EOF) {
+			return nil, nil
+		}
+		return nil, fmt.Errorf("read message failed: %w", err)
+	}
+	return ss, nil
+}
+
+// handler handles the status stream request/response.
+func (s *Service) handler(ctx context.Context, _ p2p.Peer, stream p2p.Stream) error {
+	loggerV2 := s.logger.V(2).Register()
+
+	w, r := protobuf.NewWriterAndReader(stream)
+	defer func() {
+		if err := stream.FullClose(); err != nil {
+			loggerV2.Debug("stream full close failed: %v", "error", err)
+		}
+	}()
+
+	var msgGet pb.Get
+	if err := r.ReadMsgWithContext(ctx, &msgGet); err != nil && !errors.Is(err, io.EOF) {
+		loggerV2.Debug("read message failed", "error", err)
+		return fmt.Errorf("read message: %w", err)
+	}
+
+	if err := w.WriteMsgWithContext(ctx, &pb.Snapshot{
+		ReserveSize:   s.reserve.ReserveSize(),
+		PullsyncRate:  s.sync.SyncRate(),
+		StorageRadius: uint32(s.radius.StorageRadius()),
+	}); err != nil {
+		loggerV2.Debug("write message failed", "error", err)
+		return fmt.Errorf("write message: %w", err)
+	}
+
+	return nil
+}
+
+// NewService creates a new status service.
+func NewService(
+	pctx context.Context,
+	logger log.Logger,
+	streamer p2p.Streamer,
+	topologyIter topology.PeerIterator,
+	reserve depthmonitor.ReserveReporter,
+	sync depthmonitor.SyncReporter,
+	radius postage.RadiusReporter,
+) *Service {
+	svc := &Service{
+		logger:       logger.WithName(loggerName).Register(),
+		streamer:     streamer,
+		topologyIter: topologyIter,
+		reserve:      reserve,
+		sync:         sync,
+		radius:       radius,
+	}
+
+	go func() {
+		peerFunc := func(address swarm.Address, po uint8) (stop, jumpToNext bool, err error) {
+			select {
+			case <-pctx.Done():
+				return true, false, nil
+			default:
+			}
+
+			ctx, cancel := context.WithTimeout(pctx, defaultSnapshotRequestTimeout)
+			defer cancel()
+
+			ss, err := svc.requestStatusSnapshot(ctx, address)
+			if err != nil {
+				svc.logger.Debug("cannot get status snapshot for peer", "peer_address", address, "error", err)
+				return false, true, nil
+			}
+
+			svc.mu.Lock()
+			svc.connectedPeersSnapshots = append(
```

Comment:
this forever appends the snapshots to the slice

Reply:
This is irrelevant since the logic was changed to on-demand.